### PR TITLE
Refactor block type

### PR DESCRIPTION
**Pull request by batzor**
Refactor `Block` type for safer getters/setters.
- change hash values to `Option<SaitoHash>` for safer memoization 
- all setters that affect the `hash` or `pre_hash` now sets the memoized hash values to `None` to make sure that outdated hash values are not used.
- `generate_hash()`, `generate_pre_hash()` will re-hash if necessary to update the memoized hash values.
- `get_hash()`, `get_pre_hash()` will try to read memoized value and will recalculate if not available (but will not update because we don't want to get mutable reference in getters. Use `generate` functions to update the hash values).

### CODE REVIEW

**Review by thaodt — COMMENTED** [review]
@batzor please check my comment. Its weird because i left comments a long time but u didnt see it.

File: src/block.rs [hunk 1]
```diff
@@ -392,6 +395,8 @@ impl Block {
 
     pub fn set_id(&mut self, id: u64) {
         self.id = id;
+        self.pre_hash = None;
```
[the remark this answers is not in the document]
Comment (clayrab) [reply src/block.rs:398]: Proper memoization. The hash is invalidated if any of these fields is changed.

[the remark this answers is not in the document]
Comment (batzor) [reply src/block.rs:398]: @thaodt Sorry, I thought its merged so didn't really check for comments. I think clay explained it already. 
I just made the block interface safer without adding any complexity outside the block module. You can just use setters and getters without being concerned if the memoized value is up-to-date.

[the remark this answers is not in the document]
Comment (batzor) [reply src/block.rs:398]: Previously, you had to rely on the programmer to memoize properly which is not really a good idea.

[the remark this answers is not in the document]
Comment (thaodt) [reply src/block.rs:398]: @clayrab @batzor nice, thank for your explanations.
All LGTM.

**Review by SankaD — APPROVED** [review]

**Pull request merged**